docker-compose.yaml

```diff
@@ -0,0 +1,35 @@
+version: "3.4"
+
+services:
+  production:
+    build: 
+      context: .
+      target: production
+      dockerfile: server/Dockerfile
+    ports: ["8000:80"]
+    environment: 
+      - LOG_LEVEL=DEBUG
+      - GUNICORN_RELOAD=True
+
+  development-server:
+    build: 
+      context: .
+      target: development
+      dockerfile: server/Dockerfile
+    volumes: ["./server:/server"]
+    ports: ["8000:80"]
+    environment: 
+      - LOG_LEVEL=DEBUG
+      - GUNICORN_RELOAD=True
+
+  development:
+    build: 
+      context: .
+      dockerfile: client/Dockerfile
+    depends_on: ["development-server"]
+    ports: ["3000:3000"]
+    volumes: ["./client:/client", "/client/node_modules"]
```

Comment:
Runs the node development server, available at `localhost:3000`. Note the two volumes; the first allows for hot reloading by having the container view the source that you're editing. The second prevents the host `node_modules` folder from mounting, meaning the `node_modules` from the `npm install` run in the docker build is preserved. Otherwise, you would not be able to run `docker-compose up development` and have it run successfully without first running `npm install` on the host machine.

Reply:
i'm still pretty fuzzy on volumes -- i've tried to do some reading on them but have not succeeded in building a good mental model yet. Any pointers?

Also is the difference in allowing and preventing just the lack of a `.`? seems overly subtle?

also what's the `client:` doing in the first volume param?

Sorry, very lost on this one.

Reply:
No problem! So, a volume mounts a directory from the _host_ machine onto a container. Remember that all paths are relative to the docker build context, which is the directory of the `Dockerfile` for a container, unless otherwise specified like we do here, to be the root of the project, which has access to both the `client` and `server` folders.

The volume syntax is `<HOST PATH>:<CONTAINER PATH>` so `"./client:/client"` mounts the relative directory `./client` from the host machine to the absolute path `/client` on the container.

The second volume I'll admit is slightly confusing to me as well, I haven't found a superb description to how these exactly work. It's called an anonymous volume. Essentially, that volume is receiving the data from the built image and mounting it to the host machine at a Docker-created unique location. What it does is it prevents the `./client:/client` volume from replacing the container's `/client/node_modules` folders, which is desired. The method is described here: https://stackoverflow.com/a/32785014